### PR TITLE
PHPCS-245 Implemented typed properties sniff

### PR DESCRIPTION
**Pull request by stefanhaelingbestit**
updated slevomant version
edited few files to resolve errors

### CODE REVIEW

**Comment by stefanhaelingbestit** [comment]
> If 0 is a valid pointer, then this update would be wrong.

Warum wurde hier die Logik von false auf -1 verändert? Damit kann die File vom phpcodesniffer in keinem anderen Handler genutzt werden, als in unserem Codesniffer.

**Comment by b3nl** [comment]
> If 0 is a valid pointer, then this update would be wrong.



> @b3nl
> 
> > If 0 is a valid pointer, then this update would be wrong.
> 
> Why was the logic changed from false to -1 here? With this our File can't be used in any other context than our codesniffer.

Don't know exactly. Not my work, i just renamed the method. I think it made the api for the first code sniffer guys more easy and now many parts of our sniffer are build on this -1 marker. Has something to do with PHPCS-153.

File: phpcs.xml.dist [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
     <rule ref="./src/Standards/BestIt/ruleset.xml" />
 
-    <config name="testVersion" value="7.1" />
+    <config name="testVersion" value="7.4"/>
```
Comment (b3nl) [inline phpcs.xml.dist:7]: Shit. Forgot this in 3.0.0

Comment (b3nl) [reply phpcs.xml.dist:7]: good eye.

File: src/Standards/BestIt/CodeSniffer/File.php [hunk 2]
```diff
@@ -12,78 +13,14 @@
...
  *
  * Wrapper Class for PhpCsFile to provide a consistent way to replace int|bool returns
- * with int returns (false => -1)
+ * with int|bool returns false
  * Additionally there could be some architecture changes in the future, like Token-Objects and so on.
  *
  * @author Nick Lubisch <nick.lubisch@bestit-online.de>
  * @package BestIt\CodeSniffer
  */
 class File extends AbstractFileDecorator
```
Comment (b3nl) [inline src/Standards/BestIt/CodeSniffer/File.php:22]: Is there any real value of this class after your changes? If not, remove it!

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php [hunk 3]
```diff
@@ -186,9 +186,9 @@ private function isSimpleText(array $possCommentToken): bool
      *
      * @param int $startPosition
      *
-     * @return int
+     * @return int|bool
```
Comment (b3nl) [inline src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:189]: Now you see, for what the decorator was used ... no type security now :(

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php [hunk 4]
```diff
@@ -346,7 +346,7 @@ private function validateOneLineSummary(): bool
         $summaryPos = $this->getSummaryPosition();
         $nextPossiblePos = $this->loadNextDocBlockContent($summaryPos);
 
-        if ($nextPossiblePos > -1) {
+        if ($nextPossiblePos !== false && $nextPossiblePos > 0) {
```
Comment (b3nl) [inline src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:349]: f*ng switch between int and bool

File: src/Standards/BestIt/ruleset.xml [hunk 5]
```diff
@@ -38,7 +38,7 @@
         </properties>
     </rule>
 
-    <rule ref="SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint">
+    <rule ref="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint">
```
Comment (b3nl) [inline src/Standards/BestIt/ruleset.xml:41]: Please overwrite it with a custom best it class and implement unit tests.

File: src/Standards/BestIt/ruleset.xml [hunk 6]
```diff
@@ -50,4 +50,8 @@
     <rule ref="Generic.Formatting.SpaceAfterCast" />
 
     <rule ref="Squiz.Strings.DoubleQuoteUsage.NotRequired" />
+
+    <rule ref="SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingAnyTypeHint">
```
Comment (b3nl) [inline src/Standards/BestIt/ruleset.xml:54]: Please overwrite it with a custom best it class and implement unit tests.

File: src/Standards/BestIt/Sniffs/Commenting/RequiredDocBlockSniff.php [hunk 7]
```diff
@@ -87,11 +92,19 @@ private function checkAndRegisterMissingDocBlock(): void
         if (!$this->getDocCommentPos()) {
             $tokenIdent = $this->getTokenName();
 
-            $exception = (new CodeError(
-                static::CODE_MISSING_DOC_BLOCK_PREFIX . ucfirst($tokenIdent),
-                self::MESSAGE_MISSING_DOC_BLOCK,
-                $this->stackPos
-            ))->setPayload([lcfirst($tokenIdent)]);
+            if ($tokenIdent == $this->registeredTokens[T_VARIABLE]) {
+                $exception = (new CodeError(
```
Comment (b3nl) [inline src/Standards/BestIt/Sniffs/Commenting/RequiredDocBlockSniff.php:96]: Please provide a unit test for this new case.

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php (hunk 3 above)
Comment (tkellner) [reply src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:189]: Yes not nice ARGH. But a decorator must have the same definitions and this is not possible in this case.

Comment (migo315) [reply src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:189]: Do you remember the discussion in the Devtreff this year where I said that the Union Type should never be used but there is still a situation where it is helpful?

Just saying ...

https://wiki.php.net/rfc/union_types_v2

File: src/Standards/BestIt/CodeSniffer/File.php [hunk 8]
```diff
@@ -4,6 +4,7 @@
 
 namespace BestIt\CodeSniffer;
 
+use http\Exception;
```
Comment (cedrix-bestit) [inline src/Standards/BestIt/CodeSniffer/File.php:7]: Can this be removed?

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php [hunk 9]
```diff
@@ -66,9 +66,9 @@ abstract class AbstractDocSniff extends AbstractSniff
     /**
      * The cached position of the summary.
      *
-     * @var int|null
+     * @var int|bool
      */
-    private $summaryPosition = -1;
+    private $summaryPosition = false;
```
Comment (cedrix-bestit) [inline src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:71]: As I understand it the purpose is to use false as an initial-value, like a helper. I dislike this use of "false" as helper for "first-use" or "undefined", whatever... Strictly speaking you have a boolean, but you only handle "false" and do not handle "true"-values. So imagine summaryPosition gets "true" then your logic breaks, even the getter-function below in line 125, which returns true, though it should only return null/int. What's the reason for not using null|int instead? Just curious... :)

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php (hunk 4 above)
Comment (cedrix-bestit) [inline src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:349]: $nextPossiblePos can only be false, 0 or positive integers, right? So how about:
if ($nextPossiblePos)

File: src/Standards/BestIt/CodeSniffer/File.php (hunk 2 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/CodeSniffer/File.php:22]: There are still methods defined in this class. I could remove the position method and another one that now would be identical to standard.

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php (hunk 4 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:349]: Changed to null/int, much nicer now

File: src/Standards/BestIt/ruleset.xml (hunk 6 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/ruleset.xml]: Done

File: src/Standards/BestIt/Sniffs/Commenting/RequiredDocBlockSniff.php (hunk 7 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/Sniffs/Commenting/RequiredDocBlockSniff.php:96]: Unittests with fixtures provided

File: src/Standards/BestIt/CodeSniffer/File.php (hunk 8 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/CodeSniffer/File.php:7]: Removed unnecessary use statements

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php (hunk 9 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:71]: Updated PR, exchanged false with null on all occasions i could find, tests run successfully so I think this works as well.

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php (hunk 4 above)
Comment (stefanhaelingbestit) [reply src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php]: Changed from bool to null/int, adjusted this as well

File: src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php (hunk 9 above)
Comment (mKnoop) [reply src/Standards/BestIt/Sniffs/Commenting/AbstractDocSniff.php:71]: I don't like this. We shouldn't introduce peroperties that can be multiple types. I see the problem that null is a valid value for the cached position and null indicates also that the cache is not filled. Without a proper caching implementation via PSR-6 i don't see a clean solution. So its okay for me

File: tests/Sniffs/Commenting/Fixtures/RequiredDocBlockSniff/correct/Variable.php [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+/**
+ * Valid Doc Block.
+ */
+class ClassWithDocBlock {
+    protected string $foo = 'BAR';
```
Comment (b3nl) [inline tests/Sniffs/Commenting/Fixtures/RequiredDocBlockSniff/correct/Variable.php:7]: This does not make any sense for the ADDED ERROR with the message self::MESSAGE_MISSING_DOC_BLOCK_VAR ... There should be a new error test and not an added correct fixture. And event then: This feature should not remove any doc blocks!

**Pull request merged**